### ISSUE TITLE
Features: Schema

### ISSUE THREAD

**Issue by scofalik** [ckeditor/ckeditor5-design#126]
# Features concept

This is a part of Features concept. This issue discusses schema. See also #129.
## Schema

Schema is a way to validate and organize elements positioning and occurrence in the Tree Model and Tree View. Generally speaking, for each element, schema says what elements can be contained in them, what attributes can be set on that element, and what attributes can be set on that element's children. Schema may be then used to verify if Tree Model or Tree View is properly built and also can be used to build a proper Tree Model or Tree View.

Due to flexible and modular architecture of CKEditor5 that is based on "plug-able" `Feature`s, which then may introduce custom elements and attributes, it is impossible to create a static, hard-coded schema that will cover all possible features.

Instead, schema will be (mostly) built on the fly, during `Feature`s initialization. We need a flexible, yet "easy to use in simple cases" mechanism for creating schema.

_(Right now it's hard to say what exactly should be registered to schema, it will be probably used both by converters and commands. On the other hand it is used differently by Tree Model and Tree View. Maybe there should be two separate ways to add rules to schema. Are there any other places that we will also want to use schema?)_

**Comment by pjasiun** [comment]
I think that schema should be smarter then CKEditor 4 DTD. It should not only tell what elements can be directly inside certain element, but also define more advanced relations rules. For example there are plenty of elements which can go inside MathML tag, but they do not need to be directly inside, or figure for image needs to contains both image and caption. These kinds of rules should be possible to defined, maybe the simples was is to allow users to define function as a rule.

**Comment by scofalik** [comment]
The question is whether we want to keep things like image and caption as seperate nodes in tree model or attributes. If these are kept as attributes it is easier to define in schema and maybe cleaner in Tree Model (think of converters).

**Comment by pjasiun** [comment]
Schema may also need to define which attribute is an inline style (like bold) and which is an elements attribute (like src for image). On the other hand it might be better to handle this only by converters, I can not figure any example where model need this information (maybe OT?).

**Comment by pjasiun** [comment]
> The question is whether we want to keep things like image and caption as seperate nodes in tree model or attributes. If these are kept as attributes it is easier to define in schema and maybe cleaner in Tree Model (think of converters).

Caption can have child nodes, so it can not be an attribute. And even if we decide that image is an attribute, there will be cases like this (for instance widget with two nested editable). Schema should be able to define the needed substructure.

**Comment by pjasiun** [comment]
Another interesting question is: who will have an information how to fix a broken part of the tree model? Some children should be wrapped, some removed, and some parts may needs more complex fixes.

**Comment by Reinmar** [comment]
> Caption can have child nodes, so it can not be an attribute. And even if we decide that image is an attribute, there will be cases like this (for instance widget with two nested editable). Schema should be able to define the needed substructure.

What's more – before the `<figcaption>` there may be more elements than just an image. When talking about the image plugin, you can perhaps say that the image is an attribute. But when talking about more generic widgets, that doesn't sound well.

**Comment by Reinmar** [comment]
> Schema may also need to define which attribute is an inline style (like bold) and which is an elements attribute (like src for image). On the other hand it might be better to handle this only by converters, I can not figure any example where model need this information (maybe OT?).

We had similar case in the selection and copying attributes to empty blocks. Once all attributes are on the block and someone starts typing, the engine must be able to tell which of these attributes were block's attributes and which are inline ones. Hence, the attributes must be somehow defined and the schema is IMO the best place for that.

**Comment by Reinmar** [comment]
> Another interesting question is: who will have an information how to fix a broken part of the tree model? Some children should be wrapped, some removed, and some parts may needs more complex fixes.

I think that features implementing those structures. I'm worried that implementing such mechanisms in features will be hard, but I have no other idea. There cannot be a single general structure fixer, as there's no single piece of code which knows everything about all enabled features.

The fun fact is – there are only two reasons why these fixers are needed – pasting and OT. In OT the problems will be much smaller, as the original source of data contained a valid data (only the conflicts between batches could generate some incorrect structures). In case of pasting you have a content which may be totally invalid in terms of your editor and it needs to be aligned to your rules... In CKEditor 4 we do it very well as the algorithms are centralised. I've got no idea what we'll be able to achieve in CKEditor 5 :D.

**Comment by pjasiun** [comment]
> It means that converters will be priorities in the following order:
> 
> specific features,
> paragraph,
> linebreak,
> allowedContent=true feature (converting everything, like the allowedContent=true setting in CKEditor 4).

I agree but this is a job for the converter, not schema.

**Comment by Reinmar** [comment]
Summary of another F2F discussion we had.

## Fixing Content

The comment https://github.com/ckeditor/ckeditor5-design/issues/126#issuecomment-185155194 describes how the converters will take care of fixing the content. The fact to notice is that all pasted and set data will be passed through converters, so they will be able implement all the fixing logic for these sources. They may use the schema to understand some relation with other features.

The second part of fixing that needs to be done is post-fixing OT scenarios. These cases are much simpler and will be handled by post-fix callbacks registered by the features as well. Those callbacks can again query the schema to understand the content better.

## Schema

Schema needs to keep the information about the structure, however, this information is a bit dynamic.

We agreed that feature will register its elements in the schema defining:

* in what other elements this element can be nested; this can be done by either specifying the type of elements (e.g. HTML 5 says: "Permitted parent elements: Any element that can contain flow elements") or element names directly,
* what attributes the element may contain,
* what is the nature of this element (pharsing, flow, root, sth.).

Notes:

* We are unsure yet how to name those groups of elements. Initially we've been using words such us "inline" and "block", but it creates close binding to how the developer expect to display such an element and there should be no binding like this. Hence, we're considering using HTML5's naming scheme with phrasing and flow content. The obvious disadvantage of this solution would be that developers will need to learn meaning of these terms... but perhaps it's worth to totally break the mental relation between DOM (representation) and data model (semantics). We also must make sure that HTML5's naming convention is adequate for us – I'm not sure about this. Still – even if we'll decide to not use HTML5's convention we could avoid using block/inline/etc names.
* We understood that the result of "permitted parent elements" is dynamic. It may depend on what attributes are applied to the element. I propose that the feature should define a function accepting an element or its attributes and returning the value of "permitted parent elements" (instead of introducing some domain specific language here).
* We understood that the result of "nature of element" may be dynamic as well. For example in HTML5 a link will be a flow element if it contains any flow element and a phrasing element if it doesn't. In our data model link will not be an element in one of those cases at all, but perhaps such feature will be needed anyway.

## Lovely Cases

We've been discussing these decisions thinking about the image feature. The idea about it is that it should be composable. There would be a BaseImage feature introducing the basics and then CaptionedImage, AlignedImage, ResponsiveImage, etc. The developer would be able to choose 0-N features so they need to combine well into one feature.

As we learned when implementing image2 for CKEditor 4 this feature is super complex despite looking that simple.

* Captioned image becomes a block element in the DOM so it should split the block in which it was nested.
* Centered image also needs to be a block in some cases.
* The feature needs to decide on what final element a specific attributes, styles and classes need to be defined. The `src` of course should always be defined on `<img>` but some additional classes on the `figure` (otherwise the image cannot be styled). See e.g.: http://dev.ckeditor.com/ticket/13200
* In CKEditor 4 you can reproduce the following TC:
  1. Initial content: `<p>x<img>x</p>`.
  2. Add a caption to the image. It becomes: `<p>x</p><figure><img></figure><p>x</p>`.
  3. Remove the caption.
    * Expected: the initial state.
    * Actual: `<p>x</p><p><img></p><p>x</p>`. You are not able to tell now whether the blocks could be merged or not.
* Some people want to keep `<img>` outside paragraphs which from the semantical POV makes some sense, because image is not a paragraph just like a list isn't (how's `<p>one para</p><img><p>next para</p>` different from `<p>one para</p><list><p>next para</p>`).
* People want to render the thing differently, both in the data and inside the editor (to make styling as easy as possible). E.g. the centering of an image could be done differently than we did it in CKE4 if you don't have to support some old browsers and existing CMSes like Drupal does this differently.

The last two scenarios are especially interesting, because they prove that a single image can be rendered as a block or inline element and that its position in the data model (whether it's inside or outside a block) really matters for the semantics.

One of the interesting questions is whether adding a caption to an image should:

* wrap it with some other element (to not confuse it with the DOM, because the image can be rendered in many ways there (see – Drupal's case), let's name it `captioned`) and add `caption` inside this element,
* add the `caption` inside the `image` element.

It's very hard to explain why (because many of the mentioned and not yet mentioned aspects count), but I'm certain that the second solution is much better. But let's try to analyse what are you actually doing when adding a caption to an image or making it aligned. You are semantically extending (adding more stuff) or defining the image (adding attributes). You do not say that in order to perform this action you create a captioned entity and put the image inside it. It's the opposite – the caption is being added to the image. So the main entity is still image.

Another thing we need to realise is what should happen to the image once it got captioned (what visually may change its nature to block, but only if such a renderer is enabled). In my opinion, in the data model it should stay inside the `paragraph` element in which it was initially (if it was there). However, the renderer (converter) needs to render it outside the `<p>` element.

Once we agree about this all starts to fit. All the cases start to work and you start noticing additional rules:

* The converters are the "view" layer of the architecture and hence should take care of the entire rendering. The data model should not be in any way affected by this.
* The converters can be changed – one converter is just one view of the model. You want to render the image differently – change the converter, nothing more.
* The rendering of a captioned image would work like this (isn't it resembling rendering of a styled text?):
  * first, the base converter gets the `image` element and renders it as `<img>`,
  * then the captioned image converter finds the `caption` element inside it does to the `<img>` what it wants (e.g. wraps it with a `<figure>`).

  You can notice now that the captioned image converter doesn't need to know much about what it's wrapping. It makes it more reusable for other content.

  You can say that if in the data model you had structure like `captioned > img + caption` you could implement a generic converter for the `captioned` element. True. But then the specific, `captioned` element converter would be called before its subject, so the subject would need to take care of aligning itself (e.g. moving some attrs) to the surrounding. Instead, it should be the `captioned` element converter who adds more flavour to the image, just like attributes add. Imagine what would happen if `captioned` element converter was meant to add a title to an image instead of rendering some `<figure>`... :D.
* Since the `image` element is the most outer entity in the data model it can guard its integrity and validity and its extensions extend also these aspects (e.g. the schema). If we started saying that the extension of that feature takes the whole responsibility it would mean that it needs to know more about other possible extensions. That could lead to serious issues.

## Summing Up

I'm sure you can add "but" to nearly every paragraph in the above post. However, some things cannot be denied, such as that the model should not reflect how its represented, because it may have many representations. Please remember about that when commenting, because if you start looking at the problem from a wrong perspective you may easily give arguments for other solutions.

How does all that affect schema? I think that it's crucial to understand how some complex features will operate on the data model and render the content to understand what the schema must allow. This post certainly doesn't answer all questions, but it should give some ground for further analysis of what we need.

**Comment by pjasiun** [comment]
So instead of "but" I will only said that if the tree model will be `<p>foo<img>bar</p>` and the tree view `<p>foo</p><img><p>bar</p>` mapping position will be far more complex that it is now and it will bring some ugly edge cases.

**Comment by Reinmar** [comment]
Hm... now's the time when I disagree with myself. I went too far with one thing in the previous message. I wrote that converters are data model's views, so that the way how the content needs to be rendered should not affect the model. That was a bit imprecise – the converters are data model elements' views. This changes things, because views should rather be independent from each other. One view should not affect its parent view – in this case the `image` element cannot split the block within which it's located.

So good news @pjasiun – I don't think any more that converters should be able to split existing blocks :D.

**Comment by Reinmar** [comment]
I've been also thinking more about the `<figure>` element and I start to hate it. The fact that it cannot be nested in a paragraph makes all this so much harder ;/. Its description kinda clarifies this though:

> The figure element represents a unit of content, optionally with a caption, that is self-contained, that is typically referenced as a single unit from the main flow of the document, and that **can be moved away** from the main flow of the document without affecting the document’s meaning.

So the idea is that a captioned thing makes no sense in the middle of some content.

I even wanted to write that presumably we'll want to implement the image feature so you can only insert images between blocks (because that's what users usually want to do). I don't know though how this can be done from the UX perspective (how to explain it to the user)... but it means that perhaps we can have inline images and block images working differently. I hate to say that but it sounds like we are talking here about two different features, so two different data model elements.

This turns around one more thing which I wrote before :D.

**Comment by Reinmar** [comment]
And the last question – what is the main entity in the captioned image? "The image" or "the captioned thing". What is more important from the data model perspective?

Drupal for instance produce a single tag for the whole captioned image or video. Like this:

```html
<img src="..." data-caption="..." data-align="...">
```

Thanks to that you can still easily display the image even if the caption or alignment feature is disabled in a specific configuration (e.g. for mobiles or print – so for different views to the same content).

On the other hand, a `<figure>` may contain multiple images (or any content in any amounts in fact). I do think however that a "captioned content" is a totally different feature. Perhaps when you're creating a webpage there's a thin line between these cases, but for content editing they will be totally different. For example alignment of an image inside a "captioned content" widget should not affect the parent. In "captioned image" feature it should, because it's a single entity.

Uh... I'm glad that I didn't have to change my opinion about this one :D.

**Comment by Reinmar** [comment]
@scofalik: Have fun figuring out how all this affect the requirements for the schema :D. Hopefully not much. A quick guess would be that "permitted parent elements" should still be dynamic, but "element type (phrasing/flow/...)" not necessarily (but why not if that's possible).

**Comment by scofalik** [comment]
I will need some time to wrap my head around those conclusion, but right now I see things like this:
* I like switching to using phrasing and flow content or any other abstract kind of grouping we find fitting, that is not connected with DOM rendering
* we should start talking about `TreeController`. We should avoid making `Schema` a bag for everything. When talking about "registerting fixing callbacks" I propose to think about `TreeController` and trying to put them there. We could also registered additional information on attributes there (i.e. whether it is styling attribute that should be copied from parent to selection)
* we should ask ourselves what are `Schema` responsibilities and how we want to use it. During our discussion we considered both very simplistic solutions and powerful ones.
* I agree that we could use functions to describe what can be a parent of an element. But if we go this route, we have to ask ourselves whether we will need something like ("centralized") `Schema` at all. At this point all it would do is: keep callbacks and group some items together. Probably it is still good to keep it to organize things together.
* I am all for "block" images and not bothering with inline images. If we ever decide that we want pasting inline images, do them as separate feature. Alternatively, I also agree to have two features right away.
* if we use callbacks to define valid parents, dynamic element "nature" (flow/phrasing) shouldn't be that much of a problem (hopefully).

**Comment by Reinmar** [comment]
> We should avoid making Schema a bag for everything. When talking about "registerting fixing callbacks" I propose to think about TreeController and trying to put them there.

+1.

> We could also registered additional information on attributes there (i.e. whether it is styling attribute that should be copied from parent to selection).

-1. The schema is a definition of the entities used in the data model. Attributes are one of those entities, so I would keep information about them in the schema as well.

> we should ask ourselves what are Schema responsibilities and how we want to use it

For me it should:

* store definitions of entities used in the content (elements, attributes),
* have simple methods for registering new definitions,
* have simple methods for querying.

Nothing more – no logic, just information (sometimes the information is dynamic, but it's still information – it does not do anything itself).

> I agree that we could use functions to describe what can be a parent of an element. But if we go this route, we have to ask ourselves whether we will need something like ("centralized") Schema at all. At this point all it would do is: keep callbacks and group some items together. Probably it is still good to keep it to organize things together.

I think that I answered why we need the schema above. We need to keep the information somewhere and have methods to work with this information.

**Comment by scofalik** [comment]
I've pushed first draft of how `Schema` might look like to branch `t/features` https://github.com/ckeditor/ckeditor5-core/tree/t/features

Unfortunately it is a bit out-dated when compared to our discussion above, but you may still look to see how it would be used. It implements:
* simple way of checking what are valid parents for Element E (optionally with attribute A) by using hard-typed element names,
* disallow rules to override more general allow rules with specific disallow rules,
* inherting - if you register element `myP` that bases on `p`, it will have and maintain same schema paths for checking what are valid parents,
* grouping (block/inline) - which is done thanks to inherting.

If we want to keep this architecture we have to fit valid-parent-callbacks in it. I could check how it would look like if `inside` property is a function callback. If that would work fine, we could think about further improvements.

Take a look into `Schema` tests and `AttributeCommand` code and tests to see how you can use `Schema` right now.

**Comment by scofalik** [comment]
One thing that I like in function callbacks vs hard-typed grammar is that with the second one you might have many many paths which may result in inefficiency inside `SchemaItem#_hasMatchingPath` https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R176

**Comment by scofalik** [comment]
Maybe if SchemaItem is named (`Schema#name` property) we could make

**Comment by Reinmar** [comment]
Some minor things:

* I wasn't initially thinking about such specialised methods in the scheme, but I think I like it – https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-f773ec75a68e511125655a5e6f348c68R34
* I think that types of elements should be prefixed with something – e.g. a `$` character like in CKEditor 4 – https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R19
* Maybe let use the name `extends` – https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R52
* Use maps, Luck – https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R16
* `_updateSelectionAttributes`? https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-b918bc1b232eb5ceb750af2f9e3bad2cR209
* Unnecessary because you create the item and it would be super obvious if it failed – https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R130

**Comment by Reinmar** [comment]
`registerItem` -> `register`, because you `allow` not `allowItem`.

**Comment by pjasiun** [comment]
> I think that types of elements should be prefixed with something – e.g. a $ character like in CKEditor 4 – https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R19

Maybe upper case, like enums.

**Comment by scofalik** [comment]
Well, I `registerItem` but allow for something :P. `setAllowRule` might be more fitting. `registerItem` instead of `registerElement` because we register `SchemaItem`s. We might register something like `$inline` but it does not mean that we have `$inline` element in Tree Model.

**Comment by pjasiun** [comment]
> _updateSelectionAttributes? https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-b918bc1b232eb5ceb750af2f9e3bad2cR209

Yep, `set*` without a parameter is a small WTF. The only valid measurement of code quality is WTF/minute.

**Comment by scofalik** [comment]
As far as this method is concerned, I was more worried about whether it should be in `Document` or in `Selection` than it's name.

**Comment by pjasiun** [comment]
https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-f773ec75a68e511125655a5e6f348c68R44 `inline` means `text`? Shoudn't `text` extends `inline`? It would be cleaner. Also I think that keywords should be all upper case.

**Comment by pjasiun** [comment]
https://github.com/ckeditor/ckeditor5-core/compare/t/features?expand=1#diff-1b21ae31202733dc0c9c48a31aa51040R21 I do not think that root should be a special type. In can be an element of any type, converted from the DOM element with `contenteditable` attribute. It the DOM element is `div` the model root element will be `div`. If the DOM root element is `h1` the model root element will be `header`. Schema rules will work fine.

Now, it does not make sense to disallow `inline` in `root`, because if `root` will be header, text can go directly into it.

**Comment by scofalik** [comment]
Updates:
* prefixed special items in `Schema` with `$`
* changed parameter name `baseOn` to `isExtending` (`extends` is resevered word AFAIR), also changed `_baseChains` to `_extensionChains`
* changed `_items` and `_extensionChains` from objects to maps. I left `hasItem` and `_getItem` to keep `_items` private
* changed `_setSelectionAttributes` to `_updateSelectionAttributes`. Still keeping it in `Document`.
* removed throwing error in `SchemaItem` constructor
* introduced `$text` special item in `Schema`.

> I do not think that root should be a special type. In can be an element of any type, converted from the DOM element with contenteditable attribute. It the DOM element is div the model root element will be div. If the DOM root element is h1 the model root element will be header. Schema rules will work fine.

I think this makes sense. `Schema` probably shouldn't register `$root` item by default, `RootElement` shouldn't always have same name and creator should take care what is the type of `RootElement` and what rule is added to `Schema`.

**Comment by Reinmar** [comment]
The first version of schema was implemented a long time ago. And since then we learned a lot about the requirements and plan to rewrite this implementation. See https://github.com/ckeditor/ckeditor5-engine/issues/532.
